File: apps/android/app/src/main/java/ai/openclaw/app/gateway/CloudflareAccessJWT.kt
package ai.openclaw.app.gateway

import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.doubleOrNull
import kotlinx.serialization.json.jsonObject
import java.math.BigInteger
import java.net.URI
import java.security.KeyFactory
import java.security.Signature
import java.security.spec.RSAPublicKeySpec
import java.util.Base64
import java.util.Locale

/** Cloudflare's signed discovery and app tokens use RS256, verified by the platform provider. */
internal object CloudflareAccessJWT {
  data class Claims(
    val subject: String,
    val expiresAt: Double,
  )

  fun issuer(authDomain: String): URI {
    val host = authDomain.lowercase(Locale.ROOT)
    if (!Regex("[a-z0-9](?:[a-z0-9-]{0,61}[a-z0-9])?\\.cloudflareaccess\\.com").matches(host)) invalid()
    return URI("https://$host")
  }

  fun application(
    token: String,
    origin: CloudflareAccessOrigin,
    now: Double = System.currentTimeMillis() / 1000.0,
  ): CloudflareAccessApplication {
    val metadata = decode(token)
    val audience = metadata.string("aud")
    val issuedAt = metadata.number("iat")
    if (
      metadata.string("type") != "match" || metadata.string("hostname").lowercase(Locale.ROOT) != origin.uri.host ||
      audience.isEmpty() || audience.length > 512 || issuedAt <= 0 || issuedAt < now - 86400 || issuedAt > now + 300
    ) {
      invalid()
    }
    return CloudflareAccessApplication(origin, issuer(metadata.string("auth_domain")), audience)
  }

  fun appClaims(
    token: String,
    application: CloudflareAccessApplication,
    now: Double = System.currentTimeMillis() / 1000.0,
  ): Claims {
    val claims = decode(token)
    val audience =
      when (val value = claims["aud"]) {
        is JsonPrimitive -> if (value.isString) listOf(value.content) else invalid()
        is JsonArray -> value.map { (it as? JsonPrimitive)?.takeIf(JsonPrimitive::isString)?.content ?: invalid() }
        else -> invalid()
      }
    val subject = claims.string("sub")
    val expiresAt = claims.number("exp")
    val notBefore = if (claims.containsKey("nbf")) claims.number("nbf") else null
    if (
      issuer(application.issuer.host.orEmpty()) != application.issuer || claims.string("iss") != application.issuer.toString() ||
      audience.size > 16 || application.audience !in audience || claims.string("type") != "app" ||
      subject.isEmpty() || subject.length > 512 || expiresAt <= now || (notBefore != null && notBefore > now)
    ) {
      invalid()
    }
    return Claims(subject, expiresAt)
  }

  fun verify(
    token: String,
    jwks: ByteArray,
  ) {
    val parts = parts(token)
    val header = parse(base64URL(parts[0]))
    val keys = parse(jwks)["keys"] as? JsonArray ?: invalid()
    if (keys.size > 64) invalid()
    val key =
      keys.mapNotNull { it as? JsonObject }.firstOrNull {
        it.stringOrNull("kid") == header.string("kid") && it.stringOrNull("kty") == "RSA" &&
          (it["alg"] == null || it.string("alg") == "RS256") && (it["use"] == null || it.string("use") == "sig")
      } ?: invalid()
    val modulus = base64URL(key.string("n"))
    val exponent = base64URL(key.string("e"))
    if (modulus.size !in 256..1024 || exponent.size !in 1..8) invalid()
    val publicKey = KeyFactory.getInstance("RSA").generatePublic(RSAPublicKeySpec(BigInteger(1, modulus), BigInteger(1, exponent)))
    val verifier = Signature.getInstance("SHA256withRSA")
    verifier.initVerify(publicKey)
    verifier.update("${parts[0]}.${parts[1]}".toByteArray(Charsets.US_ASCII))
    if (!verifier.verify(base64URL(parts[2]))) invalid()
  }

  // Decode only to select the constrained issuer, or recheck a token already verified at admission.
  private fun decode(token: String): JsonObject = parse(base64URL(parts(token)[1]))

  private fun parts(token: String): List<String> {
    if (token.length > 32768) invalid()
    val parts = token.split('.')
    if (parts.size != 3) invalid()
    val header = parse(base64URL(parts[0]))
    val critical = header["crit"]
    if (
      header.string("alg") != "RS256" || header.string("kid").isEmpty() || header.string("kid").length > 512 ||
      (critical != null && (critical !is JsonArray || critical.isNotEmpty()))
    ) {
      invalid()
    }
    return parts
  }

  private fun base64URL(value: String): ByteArray {
    if (value.isEmpty() || !value.all { it in 'A'..'Z' || it in 'a'..'z' || it in '0'..'9' || it == '-' || it == '_' }) invalid()
    return try {
      Base64.getUrlDecoder().decode(value)
    } catch (_: IllegalArgumentException) {
      invalid()
    }
  }

  private fun parse(bytes: ByteArray): JsonObject =
    try {
      Json.parseToJsonElement(bytes.decodeToString(throwOnInvalidSequence = true)).jsonObject
    } catch (_: IllegalArgumentException) {
      invalid()
    }

  private fun JsonObject.stringOrNull(key: String): String? = (this[key] as? JsonPrimitive)?.takeIf(JsonPrimitive::isString)?.content

  private fun JsonObject.string(key: String): String = stringOrNull(key) ?: invalid()

  private fun JsonObject.number(key: String): Double = (this[key] as? JsonPrimitive)?.takeUnless(JsonPrimitive::isString)?.doubleOrNull?.takeIf(Double::isFinite) ?: invalid()

  private fun invalid(): Nothing = throw CloudflareAccessException(CloudflareAccessException.Kind.InvalidSession)
}

File: apps/ios/Sources/Gateway/CloudflareAccessJWT.swift
import Foundation
import Security

/// Cloudflare's metadata and app tokens use RS256; verification stays in the platform crypto API.
enum CloudflareAccessJWT {
    struct Metadata: Decodable {
        let type: String
        let hostname: String
        let authDomain: String
        let aud: String
        let iat: Double

        enum CodingKeys: String, CodingKey {
            case type, hostname, aud, iat
            case authDomain = "auth_domain"
        }
    }

    struct Claims: Decodable {
        let iss: String
        let aud: Audience
        let type: String
        let sub: String
        let exp: Double
        let nbf: Double?
    }

    struct Audience: Decodable {
        let values: [String]

        init(from decoder: Decoder) throws {
            let container = try decoder.singleValueContainer()
            if let value = try? container.decode(String.self) {
                self.values = [value]
            } else {
                self.values = try container.decode([String].self)
            }
        }
    }

    private struct Header: Decodable {
        let alg: String
        let kid: String
        let crit: [String]?
    }

    private struct KeySet: Decodable {
        struct Key: Decodable {
            let kty: String
            let kid: String?
            let alg: String?
            let use: String?
            let n: String?
            let e: String?
        }

        let keys: [Key]
    }

    static func issuer(authDomain: String) throws -> URL {
        let host = authDomain.lowercased()
        let suffix = ".cloudflareaccess.com"
        let team = host.dropLast(suffix.count)
        guard let url = URL(string: "https://\(host)"),
              url.host == host, host.hasSuffix(suffix), (1...63).contains(team.utf8.count),
              team.first != "-", team.last != "-", team.utf8.allSatisfy({
                  (97...122).contains($0) || (48...57).contains($0) || $0 == 45
              }),
              url.port == nil, url.user == nil, url.password == nil,
              url.path.isEmpty, url.query == nil, url.fragment == nil
        else { throw CloudflareAccessError.invalidApplication }
        return url
    }

    static func application(
        metadata: Metadata,
        origin: CloudflareAccessOrigin,
        now: Date = Date()) throws -> CloudflareAccessApplication
    {
        guard metadata.type == "match", metadata.hostname.lowercased() == origin.url.host,
              !metadata.aud.isEmpty, metadata.aud.utf8.count <= 512,
              metadata.iat > 0, metadata.iat >= now.timeIntervalSince1970 - 86400,
              metadata.iat <= now.timeIntervalSince1970 + 300
        else { throw CloudflareAccessError.invalidApplication }
        return try CloudflareAccessApplication(
            origin: origin,
            issuer: self.issuer(authDomain: metadata.authDomain),
            audience: metadata.aud)
    }

    static func appClaims(
        _ token: String,
        application: CloudflareAccessApplication,
        now: Date = Date()) throws -> Claims
    {
        let claims = try self.decode(Claims.self, token: token)
        guard let host = application.issuer.host,
              try self.issuer(authDomain: host) == application.issuer,
              claims.iss == application.issuer.absoluteString,
              claims.aud.values.contains(application.audience), claims.aud.values.count <= 16,
              claims.type == "app", !claims.sub.isEmpty, claims.sub.utf8.count <= 512,
              claims.exp.isFinite, claims.exp > now.timeIntervalSince1970,
              claims.nbf.map({ $0.isFinite && $0 <= now.timeIntervalSince1970 }) ?? true
        else { throw CloudflareAccessError.invalidSession }
        return claims
    }

    /// Decoding is only for selecting the constrained issuer or checking an already verified token.
    static func decode<T: Decodable>(_ type: T.Type, token: String) throws -> T {
        let parts = try self.parts(token)
        return try JSONDecoder().decode(type, from: self.base64URL(parts[1]))
    }

    static func verify(_ token: String, jwks: Data) throws {
        let parts = try self.parts(token)
        let header = try JSONDecoder().decode(Header.self, from: self.base64URL(parts[0]))
        let keySet = try JSONDecoder().decode(KeySet.self, from: jwks)
        guard keySet.keys.count <= 64,
              let key = keySet.keys.first(where: {
                  $0.kid == header.kid && $0.kty == "RSA"
                      && ($0.alg == nil || $0.alg == "RS256") && ($0.use == nil || $0.use == "sig")
              }),
              let modulus = key.n, let exponent = key.e
        else { throw CloudflareAccessError.invalidSession }
        let n = try self.base64URL(Substring(modulus))
        let e = try self.base64URL(Substring(exponent))
        guard (256...1024).contains(n.count), (1...8).contains(e.count) else {
            throw CloudflareAccessError.invalidSession
        }
        let der = self.der(tag: 0x30, data: self.integer(n) + self.integer(e))
        guard let publicKey = SecKeyCreateWithData(
            der as CFData,
            [
                kSecAttrKeyType as String: kSecAttrKeyTypeRSA,
                kSecAttrKeyClass as String: kSecAttrKeyClassPublic,
            ] as CFDictionary,
            nil),
            try SecKeyVerifySignature(
                publicKey,
                .rsaSignatureMessagePKCS1v15SHA256,
                Data("\(parts[0]).\(parts[1])".utf8) as CFData,
                self.base64URL(parts[2]) as CFData,
                nil)
        else { throw CloudflareAccessError.invalidSession }
    }

    static func base64URL(_ value: Substring) throws -> Data {
        guard !value.isEmpty, value.utf8.allSatisfy({ byte in
            (65...90).contains(byte) || (97...122).contains(byte) || (48...57).contains(byte)
                || byte == 45 || byte == 95
        }) else { throw CloudflareAccessError.invalidSession }
        let base64 = value.replacingOccurrences(of: "-", with: "+").replacingOccurrences(of: "_", with: "/")
        guard let data = Data(base64Encoded: base64 + String(repeating: "=", count: (4 - base64.count % 4) % 4)) else {
            throw CloudflareAccessError.invalidSession
        }
        return data
    }

    private static func parts(_ token: String) throws -> [Substring] {
        guard token.utf8.count <= 32768 else { throw CloudflareAccessError.invalidSession }
        let parts = token.split(separator: ".", omittingEmptySubsequences: false)
        guard parts.count == 3,
              let header = try? JSONDecoder().decode(Header.self, from: self.base64URL(parts[0])),
              header.alg == "RS256", !header.kid.isEmpty, header.kid.utf8.count <= 512,
              header.crit?.isEmpty ?? true
        else { throw CloudflareAccessError.invalidSession }
        return parts
    }

    private static func integer(_ data: Data) -> Data {
        self.der(tag: 0x02, data: (data.first.map { $0 >= 0x80 } ?? false) ? Data([0]) + data : data)
    }

    private static func der(tag: UInt8, data: Data) -> Data {
        let length = data.count < 128
            ? [UInt8(data.count)]
            : (data.count < 256 ? [0x81, UInt8(data.count)] : [0x82, UInt8(data.count >> 8), UInt8(data.count & 255)])
        return Data([tag] + length) + data
    }
}
